src/tests/ftest/cart/dual_iface_server.c

```diff
@@ -0,0 +1,434 @@
+			assert(0);
+		}
+	}
+
+	rc = crt_rank_uri_get(grp, my_rank, 0, &my_uri);
+	if (rc != 0) {
+		D_ERROR("crt_rank_uri_get() failed; rc=%d\n", rc);
+		assert(0);
+	}
+
+	rc = crt_proto_register(&my_proto_fmt);
+	if (rc != 0) {
+		D_ERROR("crt_proto_register() failed; rc=%d\n", rc);
+		assert(0);
+	}
+
+	DBG_PRINT("my_rank=%d uri=%s\n", my_rank, my_uri);
+
+	/* Write self uri into a file */
+	rc = write(fd_write, my_uri, strlen(my_uri) + 1);
+	if (rc <= 0) {
+		D_ERROR("Failed to write uri to a file\n");
+		assert(0);
+	}
+	D_FREE(my_uri);
+
+	syncfs(fd_write);
+
+	/* Give time for both servers to write to local tmp file */
+	sleep(1);
+
+	/* Read each others URIs from the file */
+	memset(other_server_uri, 0x0, MAX_URI);
+	lseek(fd_read, 0, SEEK_SET);
+	rc = read(fd_read, other_server_uri, MAX_URI);
+	if (rc < 0) {
+		perror("Failed to read ");
+		D_ERROR("Failed to read uri from a file\n");
+		assert(0);
+	}
+
+	DBG_PRINT("Other servers uri is '%s'\n", other_server_uri);
+
+	if (my_rank == 0)
+		other_rank = 1;
+	else
+		other_rank = 0;
+
+	rc = crt_group_primary_rank_add(crt_ctx[0], grp, other_rank,
+					other_server_uri);
+	if (rc != 0) {
+		D_ERROR("Failed to add rank=%d uri='%s'\n",
+			other_rank, other_server_uri);
+		assert(0);
+	}
+
+	server_ep.ep_rank = other_rank;
+	server_ep.ep_tag = tag;
+	server_ep.ep_grp = grp;
+
+	rc = crt_req_create(crt_ctx[0], &server_ep, RPC_PING, &rpc);
+	if (rc != 0) {
+		D_ERROR("crt_req_create() failed; rc=%d\n", rc);
+		assert(0);
+	}
+
+	input = crt_req_get(rpc);
+	input->src_rank = my_rank;
+	input->dst_tag = tag;
+
+	rc = crt_req_send(rpc, rpc_handle_reply, &sem);
+	if (rc != 0) {
+		D_ERROR("Failed to send rpc; rc=%d\n", rc);
+		assert(0);
+	}
+
+	tc_sem_timedwait(&sem, 10, __LINE__);
+	DBG_PRINT("Ping successful to rank=%d tag=%d\n", other_rank, tag);
+
+	/* Shutdown */
+	server_ep.ep_rank = other_rank;
+	server_ep.ep_tag = tag;
+	server_ep.ep_grp = grp;
+
+	rc = crt_req_create(crt_ctx[0], &server_ep, RPC_SHUTDOWN, &rpc);
+	if (rc != 0) {
+		D_ERROR("crt_req_create() failed; rc=%d\n", rc);
+		assert(0);
+	}
+
+	input = crt_req_get(rpc);
+	input->src_rank = my_rank;
+	input->dst_tag = tag;
+
+	rc = crt_req_send(rpc, rpc_handle_reply, &sem);
+	if (rc != 0) {
+		D_ERROR("Failed to send rpc; rc=%d\n", rc);
+		assert(0);
+	}
+
+	tc_sem_timedwait(&sem, 10, __LINE__);
+
+	/* Wait until shutdown is issued and progress threads exit */
+	for (i = 0; i < NUM_SERVER_CTX; i++)
+		pthread_join(progress_thread[i], NULL);
+
+	rc = crt_finalize();
+	if (rc != 0) {
+		D_ERROR("crt_finalize() failed with rc=%d\n", rc);
+		assert(0);
+	}
+
+	d_log_fini();
+
+	return rc;
+}
+
+static void
+print_usage(const char *msg)
+{
+	printf("Error: %s\n", msg);
+	printf("Usage: ./dual_iface_server [-i 'iface0,iface1' ");
+	printf("-d 'domain0,domain1' -p 'provider'\n");
+}
+
+int main(int argc, char **argv)
+{
+	int	pid;
+	int	fd0, fd1;
+	char	tmp_file0[] = "/tmp/server0-XXXXXX";
+	char	tmp_file1[] = "/tmp/server1-XXXXXX";
```

Comment:
Is it OK to hard code these path/file names?

Reply:
it is; we assume /tmp/ exists and -XXXXXX part will be auto-replaced by mkstemp()